feat(opinions): use llm to get correct case name for recap_into_opinions

quevon24 opened the pull request:
## Summary
<!-- What does this fix, how did you fix it, what approach did you take, what gotchas are there in your code or compromises did you make? -->

This PR adds the classify_opinion_case_name_task, which uses the LLM to extract the correct case name from the opinion content when possible. The task uses prompt templates stored in .txt files for the system and user prompts. Having the prompts in plain text files makes it easy to update and keep them separate from the code.

In the future, this design enables us to add placeholders inside these .txt prompt templates and dynamically replace them with context specific data before sending the prompt to the LLM. This will simplify prompt customization and enhance maintainability without modifying code logic.

Also included are retry logic for common OpenAI errors, safe database updates, and Sentry logging for better tracking of issues.

A few tests were added to cover normal cases, retries on rate limits.

## Deployment

**This PR should:**
<!-- The following labels control the deployment of this PR if they’re applied. -->
<!-- Please put an "X" in the box on ones that apply. -->
<!-- For more details on what pods are affected by each label, see the wiki -->
<!-- https://github.com/freelawproject/courtlistener/wiki/Pull-requests-%60skip%E2%80%90%7Btype%7D%E2%80%90deploy%60-labels -->

<!-- Check here if the entire deployment can be skipped -->
<!-- This might be the case for a small fix, a tweak to documentation or something like that. -->
- [ ] `skip-deploy` (skips everything below)
    <!-- Check here if the web tier can be skipped -->
    <!-- This is the case if you're working on code that doesn't affect the front end, like management commands, tasks, or documentation. -->
    - [ ] `skip-web-deploy`
    <!-- Check here if the deployment to celery can be skipped -->
    <!--This is the case if you make no changes to tasks.py or the code that tasks rely on. -->
    - [ ] `skip-celery-deploy`
    <!-- check this if deployment to cron jobs can be skipped -->
    <!-- This is the case if no changes are made that affect cronjobs. -->
    - [x] `skip-cronjob-deploy`
    <!-- Deployment of daemons can be skipped -->
    <!-- This is the case if you haven't updated daemons or the code they depend on. -->
    - [x] `skip-daemon-deploy`

<!-- **If deployment is required:** -->
<!-- What extra steps are needed to deploy this beyond the standard deploy? -->
<!-- Do scripts need to be run or things like that? -->
<!-- If this is more than a quick thing, a new issue should be created in our infra repo: https://github.com/freelawproject/infrastructure/issues/new (if you don’t have access to it, just put the steps here) -->
<!-- Please use an ordered list or delete this if no special steps are required: -->



<!-- Thank you for contributing and filling out this form! -->

mlissner commented:
Since this is our first PR quite like this, I think it'd be smart to have @rachlllg at least take a look. Do you mind giving a skim, Rachel?

mlissner commented on cl/corpus_importer/tasks.py:
```diff
@@ -3053,10 +3062,128 @@ def recap_document_into_opinions(
+    :param recap_document_id: RECAPDocument id
+    """
+
+    system_prompt_path = (
+        "cl/corpus_importer/prompts/case_name_extract_system.txt"
+    )
+    user_prompt_path = "cl/corpus_importer/prompts/case_name_extract_user.txt"
```
We should have a better way to import prompts. Notionally, it could be something like:

```
sys_prompt, user_prompt = get_prompts("case_name_extractor")
```

Or maybe:
```
sys_prompt = get_prompt("case_name_extractor", type="system")
user_prompt = get_prompt("case_name_extractor", type="user")
```

Not sure which is right or something else, but both are better than big long paths in our code. 

And, just stylistically, we should always use the `Path` module for this kind of thing too. If you ever see something like `some/path/to/a/file` it's a bug.

mlissner commented on cl/lib/llm.py:
```````diff
@@ -0,0 +1,68 @@
+    return _PROMPT_CACHE[prompt_path]
+
+
+def parse_llm_json(s: str) -> dict | None:
+    """
+    Parse JSON string returned by LLM, handling common formatting issues
+
+    - Removes markdown code blocks (``````json).
+    - Strips whitespace.
+    - Only catches JSON decoding errors, not all exceptions.
+
+    :param s: JSON string, possibly wrapped with markdown code blocks.
+    :return: Parsed dictionary or None if parsing fails.
+    """
+    s_clean = s.replace("``````", "").replace("``````", "").strip()
+    try:
+        return json.loads(s_clean)
+    except json.JSONDecodeError:
+        return None
```````
I thought there was a way to force JSON from the LLM? Maybe @rachlllg has ideas here too?

quevon24 commented on cl/corpus_importer/tasks.py:
```diff
@@ -3053,10 +3062,128 @@ def recap_document_into_opinions(
+        or obj.case_name_full != case_name_full
+        or obj.case_name != case_name
+    )
+
```
This is in case that for some reason the case_name_match value of the llm is incorrect?

quevon24 commented on cl/corpus_importer/tasks.py:
```diff
@@ -3053,10 +3062,128 @@ def recap_document_into_opinions(
+            logger.info(
+                "Case names successfully updated https://www.courtlistener.com/opinion/%s/decision/",
+                obj.id,
+            )
```
The LLM response is sent along with the logger message to sentry

quevon24 commented:
These are the changes I made:

- Refactored the existing `call_llm` function to use the Instructor library, this enables support for multiple llm providers beyond openai.
- Added parameters for `temperature` and `max_completion_tokens` with defaults
- Allowed passing a Pydantic response_model to parse and validate responses, returning structured typed models directly.
- Moved prompt templates from .txt files into Python variables stored in dedicated modules: `cl/corpus_importer/prompts/system.py` and `cl/corpus_importer/prompts/user.py`
- Defined CaseNameExtractionResponse Pydantic model with detailed fields matching expected LLM json output
- Updated task classify_case_name_by_llm to use instructor implementation in call_llm function
- Updated tests

quevon24 commented on cl/lib/llm.py:
```diff
@@ -0,0 +1,46 @@
+        max_tokens=max_completion_tokens,
+        response_model=response_model,
+    )
+    return response
```
Yes, with OpenAI models, instructor preserves the .usage field so you can log prompt_tokens, completion_tokens, and total_tokens directly.

For other providers:

- Some (Anthropic, Groq) expose usage via .usage or headers.
- Others (Gemini, Ollama) don’t, so you’d need to count tokens manually with a tokenizer.

rachlllg replied:
ok, I should have clarified, so for this PR, we don't think it's necessary to log and track the usage?

quevon24 commented on cl/corpus_importer/tasks.py:
```diff
@@ -3053,10 +3066,144 @@ def recap_document_into_opinions(
+        return
+
+    sub_opinion = obj.sub_opinions.all().first()
+    if not sub_opinion or not sub_opinion.plain_text:
```
This implementation is designed for recpa documents that are opinions. These come in pdf files, so it's necessary to extract the text (plain text or OCR). In this specific case, the extracted text is always stored in the plain_text field. You can see it here: https://github.com/freelawproject/courtlistener/blob/main/cl/corpus_importer/tasks.py#L3044

quevon24 commented on cl/corpus_importer/tasks.py:
```diff
@@ -3053,10 +3066,144 @@ def recap_document_into_opinions(
+
+    if not isinstance(llm_response, CaseNameExtractionResponse):
+        # Added this to avoid mypy errors
+        logger.error("Invalid llm_response type: %s", type(llm_response))
```
It's a good idea, the problem is that when using instructor and pydantic is that you can't access the raw response, this would only fail if not all the values ​​of the CaseNameExtractionResponse model are returned and I think it would be strange, also if the response is wrong you would get a ValidationError in L3139, more than anything this was added to avoid mypy errors that cannot infer the type of the llm_response variable

grossir commented on cl/corpus_importer/tasks.py:
```diff
@@ -3053,10 +3066,128 @@ def recap_document_into_opinions(
+                # Update only necessary fields
+                setattr(obj, field_name, new_case_name)
+                # Check for overlap between cluster case name and extracted case name
+                overlap = winnow_case_name(new_case_name) & winnow_case_name(
```
you could compute the `winnow_case_name(cluster_name)` only once outside of the loop

grossir commented on cl/corpus_importer/tasks.py:
```diff
@@ -3053,10 +3066,128 @@ def recap_document_into_opinions(
+        if new not in (None, "") and old != new
+    }
+
+    if changed_fields:
```
could reverse and save an identation level

```python
if not changed_fields:
      return
```

grossir reviewed (approved):
Looks good, just a couple minor comments
